base/float.jl

```diff
@@ -494,6 +494,7 @@ See also: [`iszero`](@ref), [`isone`](@ref), [`isinf`](@ref), [`ismissing`](@ref
 isnan(x::AbstractFloat) = (x != x)::Bool
 isnan(x::Number) = false
 
+isfinite(x::IEEEFloat) = x - x === zero(x)
```

Comment:
As I said in https://github.com/JuliaLang/julia/pull/46163#discussion_r928887091, I feel like this should be `iszero(x - x)`, and also `iszero` can be optimised for `IEEFloat` with `x === zero(x)`, instead of `x == zero(x)`

Reply:
```
julia> iszero(-0.0)
true
```

Reply:
so that would break the optimization. The reason for this change is that `===` on floating points can turn into a bit-compare rather than a more expensive floating point compare.

Reply:
Right, sad :confused:

Reply:
Instead then:
```
julia> Base.iszero(x::Float64) = reinterpret(UInt64, x) & 0x7fff_ffff_ffff_ffff === UInt(0)
```
(btime cannot distinguish these three approaches in performance, and returns 3.5 ns for all of them)

Reply:
I'll define the initially-proposed solution
`isfinite_1(x::IEEEFloat) = x-x===zero(x)`.
Another contender, related to the `iszero(x-x)` variant but using the other half of the fact that `x-x` can only take the values `+0.0` or `NaN`, is
`isfinite_2(x::IEEEFloat) = !isnan(x-x)`.
Of course, now that we're limited to `IEEEFloat` we can instead reach for bit-twiddling
`isfinite_3(x::IEEEFloat) = (reinterpret(Unsigned,x) & Base.exponent_mask(typeof(x))) != Base.exponent_mask(typeof(x))`
```
julia> code_native(isfinite_1,(Float64,);debuginfo=:none)
        vsubsd  %xmm0, %xmm0, %xmm0
        vmovq   %xmm0, %rax
        testq   %rax, %rax
        sete    %al
        retq

julia> code_native(isfinite_2,(Float64,);debuginfo=:none)
        vsubsd  %xmm0, %xmm0, %xmm0
        vucomisd        %xmm0, %xmm0
        setnp   %al
        retq

julia> code_native(isfinite_3,(Float64,);debuginfo=:none)
        vmovq   %xmm0, %rax
        movabsq $9218868437227405312, %rcx      # imm = 0x7FF0000000000000
        andnq   %rcx, %rax, %rax
        setne   %al
        retq
```
Compared to `isfinite_1`, `isfinite_2` is one instruction shorter in isolation (on my x86). `isfinite_3` is the same number of instructions as `isfinite_1` but one of those is a hoistable `movabsq`. Another advantage is that it requires no floating point operations. Note that, after inlining, this may not be what these functions look like in the wild.

Keep in mind that `mytest!` is not a very canonical use of `isfinite`. Applying a `@code_native` shows this fact. That said, I see the `_1` and `_2` variants benchmarking identically and `_3` about 15% faster.